### ISSUE TITLE
Additional information on /booster-catalog/booster API

### ISSUE THREAD

**Issue by gastaldi** [fabric8-launcher/launcher-planning#107]
_From @gorkem on February 8, 2018 14:58_

Che requires additional information on a booster in order to generate workspaces. We can use the `/booster-catalog/booster` to retrieve some of this information however we need the information captured on `.openshiftio/application.yml` and [configmap.properties](https://github.com/fabric8-launcher/launcher-documentation/blob/master/docs/topics/readme/cd-configmap-vert.x.properties) to be provided via the API.

_Copied from original issue: fabric8-launcher/launcher-backend#68_

**Comment by gastaldi** [comment]
@gorkem can you summarize what you need specifically from these files in this issue? This may probably be a hint for us to improve the boosters' metadata in the booster-catalog

**Comment by benoitf** [comment]
I would say we are requiring to create a workspace
- projects git URL to clone
- Images to use (with their port exposed (+description of the port)  (provided by application.yml)
- commands to run (provided by properties file)

**Comment by gastaldi** [comment]
We only have this information when the user creates a quickstart, then Launcher is not used anymore. Can we somehow push this information to your service instead?

**Comment by gastaldi** [comment]
Probably not related, but I see the https://github.com/redhat-developer/rh-che/blob/master/assembly/fabric8-stacks/src/main/resources/stacks.json already has the commands to run, maybe we can reuse them somehow? 

Also, we have an issue to [improve our CheStackDetector](https://github.com/fabric8-launcher/launcher-backend/issues/66), but it would be really nice if it were a library we could reuse in our backend. 

What do you guys think?

**Comment by gorkem** [comment]
@gastaldi We are trying to avoid the use of `stacks.json` as it is static information that creates a discrepancy between launcher and che workspace. The goal is to use the same images. 

What exactly  is the role of `CheStackDetector`?

**Comment by gastaldi** [comment]
Just had a thought that perhaps we could introduce a webhook mechanism in our launcher that would let che and other services aware when a project is launched. Need to discuss this with the team, but that would allow integration lf external services (like che) easier

**Comment by quintesse** [comment]
@gorkem 

> What exactly is the role of CheStackDetector?

It takes a look at the code in the Booster and decided what `che.fabric8.io` element to add to the OpenShift BuildConfig.

With regards to the information that Che needs I really would like to know how you imagine that information to be passed? @gastaldi is suggesting that after the Booster gets launched into the user's GitHub and on OpenShift we could call a configurable REST API with information about the Booster (the URL to the GitHub repository for example, from there you could do any kind of analysis yourself). But I'm wondering if what you need is more "static", just like the `che.fabric8.io` element added to the  BuildConfig mentioned above?

**Comment by gorkem** [comment]
One of the scenarios that we want to support is adding a badge or button somewhere that would open a Che workspace with a project generated possibly into README generated by launcher. So the metadata is not necessarily passed synchronously to Che. 

I think we want to move away from that static data. The goal is to receive the metadata of the booster and generate a workspace that tailors the metadata. At the moment we basically map `che.fabric8.io` value to an entry on the stacks.json. And that mapping is limited to what OSIO can support which is a fraction of what Che and launcher supports. Also the stacks.json and booster changes are not necessarily synchronized, leading to issues. Not to mention that this now requires additional work on Che for adding/working with new boosters.

If we follow the REST API path, I think all we need is the information that REST API needs to give the information mentioned on the description.

**Comment by gastaldi** [comment]
@quintesse what if we add as `metadata` info to the booster catalog and pass them along in this webhook? I don't like the idea of reading files specific to the README generator (as that won't work for boosters that do not have documentation -
 community provided boosters)

**Comment by gastaldi** [comment]
The booster would have something like: 
```yaml
metadata:
   che:
      command: mvn vertx:run
      ports:
         - 8080
         - 8089
```

**Comment by quintesse** [comment]
@gastaldi we could, but I like even less that we would have to sync very technical information that could very easily get out of sync with the original source (remember that even the application.yaml gets generated!)

**Comment by gastaldi** [comment]
But how often will that information change? It is something we could add in common.yamls in the new structure afaik?

**Comment by quintesse** [comment]
Wdym? Afaik the application.yaml can be different with each booster. I don't think you can depend on that being the same for each booster at all. And that it doesn't change often is not a good reason in this case, because that information would now be in 3(!) different places: the `src/main/fabric8`, the `application.yaml` and now also in the `booster.yaml` that's in a completely different repository? It will work but IMO it's just asking for trouble, when things get out of sync users will get bugs that will be very hard to find.

I also think you have to take into account that it's very likely that a webhook will not be enough for many use cases, I mean what happens when they want to make changes to the user's repository? (Add that badge to a README for example) They'll need things like GitHub accounts and user identities etc. All the things we already have inside the Launcher but things you won't have if you're just a REST endpoint somewhere outside of it.

I'm just thinking that having a way for them(us?) to just easily define extra steps that get executed at launch is a better way to deal with this.  The `OsioMissionControl.launch()` would just do whatever is necessary to set up a Che environment (or better, gather the information necessary to do so at a later time and store that information somewhere that works for them)

But it's late and I'm perhaps not making sense heh

**Comment by gastaldi** [comment]
Yeah, the README changes can be implemented in a `ProjectilePreparer`. Maybe that could work for now, but I think we should be prepared for other future integrations, not sure if putting every single specific feature that IMHO belongs to other system (che in this case) in the osio addon is a good idea. 

BTW In OSIO we can pass the OSIO token along with the request and the target system can lookup the Github credentials using that token.

**Comment by quintesse** [comment]
> not sure if putting every single specific feature that IMHO belongs to other system (che in this case) in the osio addon is a good idea.

No, perhaps not, but Che is a pretty essential part of the OSIO experience. So in this case it's a good match I think.

**Comment by gorkem** [comment]
Does that mean it would be an OSIO specific solution? That would not be enough, we have a requirement to use launcher and Che that is installed on-prem.

**Comment by quintesse** [comment]
@gorkem Ah I didn't know that. But no, it certainly wouldn't be OSIO-only. If this can be made part of the Launcher too than I'm sure we'll be happy to include it. Although I'm guessing this would mean implementing something to be able to specify on what kinds of clusters that feature can be enabled? (Guessing here that Che wouldn't ever work on starter clusters for example)

**Comment by gastaldi** [comment]
@gorkem here is a sample of our booster-catalog service and endpoints support returning metadata:

```bash
$  http http://localhost:8080/api/booster-catalog/booster\?mission\=rest-http\&runtime\=vert.x --timeout 1000

HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: *
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS, HEAD
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: *
Access-Control-Max-Age: 151200
Connection: keep-alive
Content-Length: 230
Content-Type: application/json
Date: Thu, 22 Mar 2018 19:50:18 GMT

{
    "gitRef": "master",
    "gitRepo": "https://github.com/openshiftio-vertx-boosters/vertx-http-booster",
    "id": "vert.x_community_rest-http_booster",
    "metadata": {
        "worksWithLegacyOsio": true
    },
    "mission": {
        "metadata": {
            "foo": "bar"
        }
    },
    "runtime": {
        "metadata": {
            "pipelinePlatform": "maven"
        }
    }
}
```

To add more information about it (like che parameters), you just need to add information in the booster-catalog for:
a) a specific booster - change the `metadata` in the booster.yaml
b) a specific runtime - change the `metadata` tag for the specific runtime in `metadata.yaml`
c) a specific mission - change the `metadata` tag for the specific mission in `metadata.yaml`

Let me know if you need any help

**Comment by gastaldi** [comment]
`/api/booster-catalog` already provides this information. 
See the endpoint documentation in https://editor.swagger.io/?url=https://forge.api.prod-preview.openshift.io/swagger.yaml for more details

Closing
